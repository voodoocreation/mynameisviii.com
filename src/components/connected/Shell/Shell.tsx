import cn from "classnames";
import Head from "next/head";
import * as React from "react";
import { FormattedMessage, InjectedIntl, injectIntl } from "react-intl";
import { connect } from "react-redux";

import { isServer } from "../../../helpers/dom";
import Banner from "../../presentation/Banner/Banner";
import Button from "../../presentation/Button/Button";
import Loader from "../../presentation/Loader/Loader";
import OnlineStatusToast from "../../presentation/OnlineStatusToast/OnlineStatusToast";
import Toast from "../../presentation/Toast/Toast";
import ToastContainer from "../../presentation/ToastContainer/ToastContainer";

import * as selectors from "../../../selectors/root.selectors";

interface IStoreProps {
  hasNewVersion: boolean;
  isLoading: boolean;
  isOnline: boolean;
}

interface IProps extends IStoreProps {
  className?: string;
  intl: InjectedIntl;
}

class Shell extends React.Component<IProps> {
  public componentWillMount() {
    if (!isServer()) {
      const html = document.documentElement as HTMLHtmlElement;
      html.classList.add("isClientRendered");
    }
  }

  public render() {
    const {
      children,
      className,
      hasNewVersion,
      isLoading,
      isOnline
    } = this.props;
    const { formatMessage } = this.props.intl;

    return (
      <article className={cn("Page", { isLoading }, className)}>
        <Head>
          <meta
            property="og:site_name"
            content={formatMessage({ id: "BRAND_NAME" })}
          />
        </Head>

        <Banner />

        <main className="Page-body" role="main">
          {isLoading ? <Loader className="PageLoader" /> : children}
        </main>

        <ToastContainer>
          <OnlineStatusToast isOnline={isOnline} />

          <Toast
            className="HasNewVersionToast"
            hasAutoDismiss={false}
            isVisible={hasNewVersion}
          >
            <FormattedMessage id="NEW_VERSION_AVAILABLE" />
            <Button
              className="HasNewVersionToast-refreshButton"
              onClick={this.onRefreshClick}
            >
              <FormattedMessage id="REFRESH" />
            </Button>
          </Toast>
        </ToastContainer>
      </article>
    );
  }

  private onRefreshClick = () => {
    window.location.reload(true);
  };
}

const mapStateToProps = (state: any) => ({
  hasNewVersion: selectors.hasNewVersion(state),
  isLoading: selectors.getPageIsLoading(state),
  isOnline: selectors.isOnline(state)
});

export default injectIntl<any>(connect<IStoreProps>(mapStateToProps)(Shell));
