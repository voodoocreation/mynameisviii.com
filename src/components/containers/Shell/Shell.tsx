import cn from "classnames";
import Head from "next/head";
import * as React from "react";
import { InjectedIntl, injectIntl } from "react-intl";
import { connect } from "react-redux";

import Banner from "../../presentation/Banner/Banner";

import * as selectors from "../../../selectors/root.selectors";
import Loader from "../../presentation/Loader/Loader";

interface IStoreProps {
  isLoading: boolean;
}

interface IProps extends IStoreProps {
  className?: string;
  intl: InjectedIntl;
}

class Shell extends React.Component<IProps> {
  public componentWillMount() {
    if (
      typeof document !== "undefined" &&
      !document.documentElement.classList.contains("isClientRendered")
    ) {
      document.documentElement.classList.add("isClientRendered");
    }
  }

  public render() {
    const { children, className, isLoading } = this.props;
    const { formatMessage } = this.props.intl;

    return (
      <article className={cn("Page", { isLoading }, className)}>
        <Head>
          <meta
            property="og:site_name"
            content={formatMessage({ id: "BRAND_NAME" })}
          />
        </Head>

        <Banner />

        <main className="Page-body" role="main">
          {isLoading ? <Loader className="PageLoader" /> : children}
        </main>
      </article>
    );
  }
}

const mapStateToProps = (state: any) => ({
  isLoading: selectors.getPageIsLoading(state)
});

export default injectIntl<any>(connect<IStoreProps>(mapStateToProps)(Shell));
