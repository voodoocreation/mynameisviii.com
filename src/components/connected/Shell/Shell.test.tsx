import { mount, render } from "enzyme";
import merge from "lodash.merge";
import * as React from "react";
import { Provider } from "react-redux";

import createStore from "../../../store/root.store";
import Shell from "./Shell";

const setup = (fn: any, fromTestProps?: any, fromTestStore?: any) => {
  const props = {
    className: "TestClassName",
    ...fromTestProps
  };
  const store = createStore(
    merge({
      page: { isLoading: false },
      ...fromTestStore
    })
  );

  return {
    actual: fn(
      <Provider store={store}>
        <Shell {...props} />
      </Provider>
    ),
    props,
    store
  };
};

describe("[connected] <Shell />", () => {
  it("renders correctly on the client", () => {
    const { actual } = setup(render);
    expect(actual).toMatchSnapshot();
  });

  it("renders correctly on the server", () => {
    window.isServer = true;
    const { actual } = setup(render);
    expect(actual).toMatchSnapshot();
  });

  it("renders correctly when loading", () => {
    const { actual } = setup(
      render,
      {},
      {
        page: { isLoading: true }
      }
    );
    expect(actual).toMatchSnapshot();
  });

  it("renders new version toast and refreshes page when refresh button is clicked", () => {
    window.location.reload = jest.fn();
    const { actual } = setup(
      mount,
      {},
      {
        page: { hasNewVersion: true }
      }
    );

    const toast = actual.find("Toast.HasNewVersionToast");
    expect(toast).toHaveLength(1);

    toast.find("Button.HasNewVersionToast-refreshButton").simulate("click");
    expect(window.location.reload).toHaveBeenCalledWith(true);
  });
});
