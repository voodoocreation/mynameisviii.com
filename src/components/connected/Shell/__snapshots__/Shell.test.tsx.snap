// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[containers] <Shell /> renders correctly 1`] = `
<article
  class="Page TestClassName"
>
  <header
    class="Banner"
    role="banner"
  >
    <a
      class="Brand"
      href="/"
    >
      <span>
        My name is Viii: a painter of frequencies and thoughts
      </span>
      <svg />
      <svg />
    </a>
    <nav
      class="Navigation"
      role="navigation"
    >
      <button>
        <span>
          Menu
        </span>
      </button>
      <ul>
        <li>
          <a
            href="/news"
          >
            <span>
              News
            </span>
          </a>
        </li>
        <li>
          <a
            href="/releases"
          >
            <span>
              Releases
            </span>
          </a>
        </li>
        <li>
          <a
            href="/galleries"
          >
            <span>
              Galleries
            </span>
          </a>
        </li>
        <li>
          <a
            href="/stems"
          >
            <span>
              Stems
            </span>
          </a>
        </li>
        <li>
          <a
            href="/appearances"
          >
            <span>
              Appearances
            </span>
          </a>
        </li>
      </ul>
    </nav>
  </header>
  <main
    class="Page-body"
    role="main"
  />
  <div
    class="ToastContainer"
  >
    <div
      class="Toast OnlineStatusToast isHidden"
    >
      <div
        class="Toast-content"
      >
        <div
          class="Toast-message"
        >
          <span>
            You are offline. Some data and external links will be unavailable until you reconnect.
          </span>
        </div>
        <button
          class="Toast-closeButton"
          title="Close"
        >
          <svg
            attr="[object Object]"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
          >
            <path
              d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
            />
          </svg>
        </button>
      </div>
    </div>
    <div
      class="Toast HasNewVersionToast isHidden"
    >
      <div
        class="Toast-content"
      >
        <div
          class="Toast-message"
        >
          <span>
            A new version of the site is available, refresh to see it.
          </span>
          <button
            class="Button HasNewVersionToast-refreshButton"
            type="button"
          >
            <span
              class="Button-content"
            >
              <span>
                Refresh
              </span>
            </span>
          </button>
        </div>
        <button
          class="Toast-closeButton"
          title="Close"
        >
          <svg
            attr="[object Object]"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
          >
            <path
              d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</article>
`;

exports[`[containers] <Shell /> renders correctly when loading 1`] = `
<article
  class="Page isLoading TestClassName"
>
  <header
    class="Banner"
    role="banner"
  >
    <a
      class="Brand"
      href="/"
    >
      <span>
        My name is Viii: a painter of frequencies and thoughts
      </span>
      <svg />
      <svg />
    </a>
    <nav
      class="Navigation"
      role="navigation"
    >
      <button>
        <span>
          Menu
        </span>
      </button>
      <ul>
        <li>
          <a
            href="/news"
          >
            <span>
              News
            </span>
          </a>
        </li>
        <li>
          <a
            href="/releases"
          >
            <span>
              Releases
            </span>
          </a>
        </li>
        <li>
          <a
            href="/galleries"
          >
            <span>
              Galleries
            </span>
          </a>
        </li>
        <li>
          <a
            href="/stems"
          >
            <span>
              Stems
            </span>
          </a>
        </li>
        <li>
          <a
            href="/appearances"
          >
            <span>
              Appearances
            </span>
          </a>
        </li>
      </ul>
    </nav>
  </header>
  <main
    class="Page-body"
    role="main"
  >
    <span
      class="Loader PageLoader"
    >
      <div
        class="Loader-animation"
      >
        <span />
        <span />
        <span />
      </div>
    </span>
  </main>
  <div
    class="ToastContainer"
  >
    <div
      class="Toast OnlineStatusToast isHidden"
    >
      <div
        class="Toast-content"
      >
        <div
          class="Toast-message"
        >
          <span>
            You are offline. Some data and external links will be unavailable until you reconnect.
          </span>
        </div>
        <button
          class="Toast-closeButton"
          title="Close"
        >
          <svg
            attr="[object Object]"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
          >
            <path
              d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
            />
          </svg>
        </button>
      </div>
    </div>
    <div
      class="Toast HasNewVersionToast isHidden"
    >
      <div
        class="Toast-content"
      >
        <div
          class="Toast-message"
        >
          <span>
            A new version of the site is available, refresh to see it.
          </span>
          <button
            class="Button HasNewVersionToast-refreshButton"
            type="button"
          >
            <span
              class="Button-content"
            >
              <span>
                Refresh
              </span>
            </span>
          </button>
        </div>
        <button
          class="Toast-closeButton"
          title="Close"
        >
          <svg
            attr="[object Object]"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
          >
            <path
              d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</article>
`;
